Rewrite docs/TriageProcess.md to be timeless and more comprehensive (#66566)
dotnet/aspnetcore: docs/TriageProcess.md

# Triage Process

This document describes how issues filed in `dotnet/aspnetcore` are triaged. It is intended as:

- A reference for the **.NET community** so you know what to expect after filing an issue.
- A guide for **engineers joining the team** on how to triage incoming issues consistently.

> **Need urgent help?** Customers needing urgent investigation should contact [Microsoft Support](https://dotnet.microsoft.com/platform/support).

> ## What we triage and when

Author: I think "what" we triage is obvious: the issues in this repo. This doc is more about "how" we triage.

As for "when", we should set an appropriate expectation for how long it will take someone to look at an issue. We want to be responsive to our community but also realistic. I think saying that we try to respond to new issues as quick as we can but that it may take a few days is reasonable.

> Every open issue without a milestone is considered untriaged. The team reviews untriaged issues regularly (typically weekly per area) and aims to make a decision quickly. Decisions are recorded as a **type label**, an **area label**, and a **milestone** — together they answer "what kind of issue is this, who owns it, and when (if ever) will we work on it?"

Author: Our team's exact work patterns evolve and change, so let's not commit to a weekly triage here. Saying that we review incoming issues regularly is sufficient.

Reviewer: I also would say "when (if ever) are we *planning* to work on it". This gives more impression that the milestone might not be a guaranteed timing (e.g, we can put issues in .NET X milestone and then later decide to push to .NET X+1 or even to the Backlog, I think)

## Triage outcomes

Every triaged issue ends up in one of these states:

| Outcome | Meaning |
| --- | --- |

> | **Closed** | Not actionable: invalid, duplicate, by-design, won't fix, can't repro, or already fixed. |

Author: What does "invalid" mean here?

| **Current release milestone** (e.g. the active `N.0` milestone) | Important enough to address in the in-flight release. Used sparingly. |

> | **Servicing milestone** (e.g. `N.0.x`) | A regression or high-impact bug in a supported release that warrants a patch. Subject to the [servicing bar](https://github.com/dotnet/aspnetcore/blob/main/docs/ReleasePlanning.md). |

Author: ```suggestion
| **Servicing milestone** (e.g. `N.0.x`) | A regression or high-impact bug in a supported release that warrants a patch. Subject to the [servicing bar](https://github.com/dotnet/aspnetcore/blob/main/docs/Servicing.md). |
```

> | **Next-release planning milestone** | Candidate for the next major .NET release; reviewed during sprint planning. |

Author: It's not clear what milestones these triage outcomes correspond to. I think the actual options are:
- Next servicing release: `N.0.x`
- Planning milestone for the current release: `.NET N Planning`
    - Issues in `.NET N Planning` will then get scheduled for specific release milestones (`N.0-previewX`) during team planning.
- Backlog

| **`Backlog`** | Tracked, but not committed to any release. Re-evaluated during release planning. |

> | **`Needs: Author Feedback`** | Waiting on the reporter for more information. May auto-close if no response. |

Author: Do we cover `Needs: Author Feedback` and the related issue interaction labels like `Needs: Repro` anywhere else in the repo docs that we can reference?

If the report depends on more information from the reporter, we apply `Needs: Author Feedback` first and triage again once we have what we need.

## How we triage by issue type

> Every triaged issue is tagged with a **type label**: `bug`, `enhancement` (feature), `Docs`, or `investigate` / `question`. The decision logic differs by type.

Author: I think we want to move from the type labels (bug, enhancement, task) to the issue type field, which can be Bug, Epic, Feature, or Task

### Bug reports

For something to be triaged as a bug, the issue should clearly describe:

1. The expected behavior.
2. The actual behavior.

> 3. Steps to reproduce, ideally a minimal repro project.

Author: Do we document our requirements for repro projects anywhere else? We typically want repros provided in a separate github repo.

Reviewer: @danroth27 Based on my personal prior experience, I think it's a lot better for repros to be zip files attached in the issue.

Separate repos can easily end up being deleted, esp for old issues. I have seen that in the past few times.

I can see the upside of having a separate repo is that the repro is easier to navigate in the browser without needing to download, unzip, ...

Author: @Youssef1313 We were advised to ask for repos in part because of the risk of [zip bombs](https://en.wikipedia.org/wiki/Zip_bomb).

4. The .NET version and OS.

Then we assess impact and severity, considering:

- **Is it a regression in a currently supported release?** Regressions are weighted heavily and are the strongest candidates for servicing.
- **Severity:** crash, data loss, security, hang, perf, or incorrect behavior with workaround?
- **Reach:** how many users are likely affected? Common scenario or a niche edge case?
- **Community signal:** 👍 reactions and substantive comments from distinct users.
- **Workaround availability:** is there a reasonable user-side fix?

Based on that:

> - **Critical / clear regression in a supported release** → servicing milestone (subject to servicing-bar approval).

Author: Is the servicing bar more formally refined anywhere else in the repo docs that we can just reference? There are other nuances to the servicing decision, like the risk of introducing regressions and avoiding API changes in servicing.

- **High-impact, reproducible** → next-release planning milestone.
- **Real but low impact, or unclear impact** → `Backlog`, awaiting more signal.

> - **Cannot reproduce or insufficient information** → `Needs: Author Feedback`; close if no response.

Author: For repros we use `Needs: Repro`.

- **Caused by user code or environment** → close with an explanation.

### Feature requests (`enhancement`)

We weigh:

- Alignment with the framework's goals and roadmap.
- Size and design complexity.
- Community demand (👍, comments, related issues, prior discussions).
- Whether the scenario can already be achieved via extension points.

Outcomes:

> - **Aligned and well-scoped** → next-release planning milestone for design review.

Author: We try to prioritize the most important and impactful feature requests for each release. There are multiple factors that go determining the priority of a feature. It's a balance between addressing key feedback, investing in key strategic directions, and satisfying Microsoft business needs. We also have to weigh-in the cost and risk of building, maintaining, and supporting the feature given the team's available resources.

- **Reasonable but not urgent** → `Backlog` to gather signal.
- **Out of scope or unlikely to ship** → close with rationale and, where possible, a pointer to alternatives.

### Investigations

> When a report could be a bug, environmental, or a misconfiguration, we apply the `investigate` label. Most investigations resolve to "user code / environment" and close; a minority surface as real bugs and re-triage as such.

Author: ```suggestion
When further investigation from our team is needed to determine the nature and impact of a reported issue, we apply the `investigate` label. Once the investigation is completed, the issue is triaged normally. Issues that don't have clear impact or severity may get closed without being investigated.
```

For investigations that require a process dump, ETW trace, or other diagnostic the reporter must provide, we set `Needs: Author Feedback`. If the reporter cannot or will not provide actionable diagnostics, the issue is closed; reopening is welcome once data is available.

### Documentation issues

> Tagged `Docs` and typically transferred to [`dotnet/AspNetCore.Docs`](https://github.com/dotnet/AspNetCore.Docs). High-traffic confusion may be addressed in the current release.

Author: Doc issues in the dotnet/aspnetcore repo track technical content gaps that need to be filled by the ASP.NET Core engineering team. The engineering team provides the technical content to add to the docs by opening an issue with the content in the dotnet/aspnetcore.docs repo. Doc issues should have type Task and have the Docs label.

### Wrong area or wrong repo

> If an issue belongs in another area or repo (e.g. `dotnet/runtime`, `dotnet/sdk`, `dotnet/AspNetCore.Docs`, `microsoft/reverse-proxy`), we transfer or re-label it during triage rather than holding it.

Author: ```suggestion
If an issue belongs in another repo (e.g. `dotnet/runtime`, `dotnet/sdk`, `dotnet/AspNetCore.Docs`, `dotnet/yarp`), we transfer it during triage.
```

## Stale-issue cleanup

Issues that have been in `Backlog` across multiple releases without gaining traction are closed during release planning. Long-lived backlog age is itself a signal that an issue isn't impactful enough to act on; closing keeps the backlog meaningful. Reopening is welcome when new information or community demand emerges.

## How decisions flow

```mermaid
flowchart TD
    A[New issue] --> B{Enough info?}
    B -- No --> C[Needs: Author Feedback]
    C -. no response .-> X[Close]
    C -. info provided .-> B
    B -- Yes --> D{Type?}
    D -- Bug --> E{Impact / regression?}
    D -- Feature --> F{Aligned with goals?}
    D -- Investigation --> G[Investigate]
    D -- Docs --> H[Transfer to docs repo]
    D -- Out of scope / by design / duplicate --> X
    E -- Critical / supported regression --> S[Servicing milestone]
    E -- High impact --> P[Next-release planning]
    E -- Low / unclear impact --> BL[Backlog]
    F -- Yes --> P
    F -- Maybe --> BL
    F -- No --> X
    G -- Real bug --> E
    G -- User / env issue --> X
    BL -. release planning .-> P
    BL -. stale across releases .-> X

> ```

Author: This diagram is way too complicated. Remove it.

## References

- [Issue Management Policies](https://github.com/dotnet/aspnetcore/blob/main/docs/IssueManagementPolicies.md) — automation and labels.
- [Release Planning](https://github.com/dotnet/aspnetcore/blob/main/docs/ReleasePlanning.md) — how releases are scoped.

> - [Servicing bar](https://github.com/dotnet/aspnetcore/blob/main/docs/ReleasePlanning.md) — what qualifies for a patch release.

Author: We should review these docs together and make sure the content is consistent and that we're not repeating content in multiple docs.

> - [Servicing bar](https://github.com/dotnet/aspnetcore/blob/main/docs/ReleasePlanning.md) — what qualifies for a patch release.

Author: ```suggestion
- [Servicing](https://github.com/dotnet/aspnetcore/blob/main/docs/Servicing.md) — Servicing bar and release process.
```

We assess impact and severity by considering:

- **Is it a regression in a currently supported release?** Regressions are weighted heavily and are the strongest candidates for servicing.
- **Severity:** crash, data loss, security, hang, perf, or incorrect behavior with a workaround?
- **Reach:** how many users are likely affected? Common scenario or a niche edge case?
- **Community signal:** 👍 reactions and substantive comments from distinct users.
- **Workaround availability.**

Outcomes:

- **Clear regression in a supported release with no good workaround** → servicing milestone, subject to the [servicing bar](https://github.com/dotnet/aspnetcore/blob/main/docs/Servicing.md) (which also weighs regression risk and forbids API changes in patches).
- **High impact, reproducible** → current-release planning milestone.
- **Real but low impact, or unclear impact** → `Backlog`.
- **Caused by user code or environment** → close with explanation.

### Feature

> We try to prioritize the most important and impactful feature requests for each release. There are multiple factors that go into determining the priority of a feature. It's a balance between addressing key feedback, investing in key strategic directions, and satisfying Microsoft business needs. We also have to weigh the cost and risk of building, maintaining, and supporting the feature given the team's available resources.

Author: ```suggestion
We try to prioritize the most important and impactful feature requests for each release. There are multiple factors that go into determining the priority of a feature. It's a balance between addressing key feedback, investing in key strategic directions, and satisfying business needs. We also have to weigh the cost and risk of building, maintaining, and supporting the feature given the team's available resources.
```

A milestone reflects current intent, not a guarantee. Issues placed in a planning milestone may be rescheduled, moved to `Backlog`, or closed as priorities evolve.

## Triage outcomes

A triaged issue ends up in one of these states:

| Outcome | Meaning |
| --- | --- |
| **Closed** | Not actionable: duplicate, by-design, won't fix, can't repro, or already fixed. |
| **Servicing milestone** (`N.0.x`) | A regression or high-impact bug in a supported release that warrants a patch. Subject to the [servicing bar](./Servicing.md). |
| **Current release milestone** (`N.0-previewX`) | Needs to be addressed in the next preview of the in-flight release — for example, a recent regression or a blocker for an in-progress feature. |
| **Current-release planning milestone** (`.NET N Planning`) | Candidate for the in-flight major release. Issues here are scheduled into a specific `N.0-previewX` milestone during team planning. |
| **`Backlog`** | Tracked but not committed to any release. Re-evaluated during release planning. |

> If we need more information from the reporter before we can triage, we apply the `Needs: Author Feedback` or `Needs: Repro` label and triage again once the information is provided. See [Issue Management Policies](./IssueManagementPolicies.md) for how those labels behave.

Author: Addressed in 312d43c — added a `Needs: Repro` subsection to `docs/IssueManagementPolicies.md` documenting the bot behavior (templated reply linking to `repro.md`, auto-applies `Needs: Author Feedback`, inheriting the 7-day auto-close).

## Investigations

When further investigation from our team is needed to determine the nature and impact of a reported issue, we apply the `investigate` label. Once the investigation is completed, the issue is triaged normally. Issues that don't have clear impact or severity may get closed without being investigated.

## Wrong repo

If an issue belongs in another repo (e.g. `dotnet/runtime`, `dotnet/sdk`, `dotnet/AspNetCore.Docs`, `dotnet/yarp`), we transfer it during triage.

## Release planning

Near the end of a release cycle, the team reviews `Backlog` and promotes items aligned with the next release into the next planning milestone. Issues that have stayed in `Backlog` across multiple releases without gaining traction are typically closed during this review — long backlog age is itself a signal that an issue isn't impactful enough to act on. Reopening is welcome when new information or community demand emerges.

See [Release Planning](./ReleasePlanning.md) for details.

## References

> - [Issue Management Policies](./IssueManagementPolicies.md) — automation and labels (`Needs: Author Feedback`, `Needs: Repro`, etc.).

Author: Same as above — addressed in 312d43c.

- [Bug Report Reproduction Guide](./repro.md) — what we expect from a minimal repro.
- [Release Planning](./ReleasePlanning.md) — how releases are scoped.

> - [Servicing](./Servicing.md) — Servicing bar and release process.

Author: Updated the PR description: the Mermaid flowchart language has been removed (the diagram was dropped during review per @danroth27's feedback).
